### PR TITLE
[RC] Implement RC trigger of ha failover of metrics/logs individually

### PR DESCRIPTION
**Pull request by moth-a**
### What does this PR do?

Updates the AGENT_FAILOVER remote config product client to support two boolean keys instead of one, which allows us to trigger the failover any combination of metrics or logs.

Additionally, this PR includes a partial refactor of the term `high availability (HA)` to `multi-region failover (MRF)` where applicable.

### What does this PR NOT do?
- It does not refactor the rest of the `ha rc client` component to properly use `mrf`.
- It does not refactor the existing `ha` setting group to be named `multi_region_failover`
- It does not introduce the behaviors that will be driven by the `multi_region_failover.failover_metrics` and `multi_region_failover.failover_logs` runtime settings.

All of these will be done separately in a PR from @tobz (#24923)

### Motivation

HAMR Q2 objective

### Additional Notes

This implementation may change soon depending on requirements for filtering logs and metrics that are emitted to R2

### Describe how to test/QA your changes

Configure an agent with 
```
api_key: ***R1_site_RC_key***
site: datad0g.com
env: staging
remote_configuration:
  enabled: true
multi_region_failover:
  enabled: true
  failover: false
  api_key: ***R2_site_RC_key***
  site: us5.datadoghq.com
```

From R2, trigger a failover targeting your agent. An example request can be found [here](https://datadoghq.atlassian.net/wiki/spaces/ENGPROG/pages/3539436450/HAMR+Gameday+Iteration+5+Runbook), but change the payload to use the two new `failover_metrics` and `failover_logs` fields instead. The agent will perform info logs as it manages the runtime settings.

Delete the failover config, and verify that the agent logs that it is unsetting the fields that were set by your config, e.g. `Falling back to `ha.failover_metrics: %t``

### CODE REVIEW

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 1]
```diff
@@ -161,48 +161,79 @@ func (rc rcClient) start() {
...
-		if haUpdate != nil && haUpdate.Failover != nil {
+
+		if haUpdate != nil && (haUpdate.FailoverMetrics != nil || haUpdate.FailoverLogs != nil) {
+			// If we've received multiple config files updating the failover settings, we should disregard all but the first update and log it, as this is unexpected
 			if applied {
-				pkglog.Infof("Multiple HA updates received, disregarding update of `ha.failover` to %t", *haUpdate.Failover)
+				pkglog.Infof("Multiple HA updates received, disregarding update of `ha.failover_metrics` to %v and `ha.failover_logs` to %v", haUpdate.FailoverMetrics, haUpdate.FailoverLogs)
+				applyStateCallback(cfgPath, state.ApplyStatus{
+					State: state.ApplyStateError,
+					Error: err.Error(),
```
Comment (mellon85) [inline comp/remote-config/rcclient/rcclientimpl/rcclient.go:203]: this is going to be the same error as above (the parsing error), we will need a new one to express this situation (so we can also monitor it easily)

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 2]
```diff
@@ -161,48 +161,79 @@ func (rc rcClient) start() {
...
+			}
+			if haUpdate.FailoverLogs != nil {
+				appliedLogs := rc.applyHARuntimeSetting("ha.failover_logs", *haUpdate.FailoverLogs, cfgPath, applyStateCallback)
+				applied = applied || appliedLogs
+			}
+			if applied {
 				applyStateCallback(cfgPath, state.ApplyStatus{
-					State: state.ApplyStateError,
+					State: state.ApplyStateAcknowledged,
 					Error: err.Error(),
```
Comment (mellon85) [inline comp/remote-config/rcclient/rcclientimpl/rcclient.go:219]: Err should be nil here

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 3]
```diff
@@ -161,60 +161,76 @@ func (rc rcClient) start() {
 }
 
 func (rc rcClient) mrfUpdateCallback(updates map[string]state.RawConfig, applyStateCallback func(string, state.ApplyStatus)) {
-	var failover *bool
+	// If the updates map is empty, we should unset the failover settings if they were set via RC previously
+	if len(updates) == 0 {
+		// Determine the current source of the multi_region_failover.failover_metrics and multi_region_failover.failover_logs cfg value
```
Comment (ameske) [inline comp/remote-config/rcclient/rcclientimpl/rcclient.go:166]: Nit: The code describes this sufficiently.

**Review by ameske — COMMENTED** [review]
Mostly just style nits, but one question about a potential situation involving one of metrics/logs failing to be set at runtime.

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go (hunk 3 above)
Comment (timothyalexandersoftware) [reply comp/remote-config/rcclient/rcclientimpl/rcclient.go]: deleted in [fbffb0a](https://github.com/DataDog/datadog-agent/pull/24754/commits/fbffb0a537975d05cdb1303171399747a2fab506) 👍

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 4]
```diff
@@ -161,60 +161,76 @@ func (rc rcClient) start() {
...
+	pkglog.Infof("Setting `%s: %t` through remote config", setting, value)
+	err := rc.settingsComponent.SetRuntimeSetting(setting, value, model.SourceRC)
+	if err != nil {
+		pkglog.Errorf("Failed to set %s runtime setting to %t: %s", setting, value, err)
+		applyStateCallback(cfgPath, state.ApplyStatus{
+			State: state.ApplyStateError,
+			Error: err.Error(),
+		})
 	}
+	return err == nil
```
[the remark this answers is not in the document]
Comment (timothyalexandersoftware) [reply comp/remote-config/rcclient/rcclientimpl/rcclient.go]: updated in [fbffb0a](https://github.com/DataDog/datadog-agent/pull/24754/commits/fbffb0a537975d05cdb1303171399747a2fab506) 👍

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 5]
```diff
@@ -161,60 +161,76 @@ func (rc rcClient) start() {
...
+		}
+		if mrfFailoverLogsSource == model.SourceRC {
+			pkglog.Infof("Falling back to `multi_region_failover.failover_logs: %t`", config.Datadog.GetBool("multi_region_failover.failover_logs"))
+		}
+		return
+	}
+
 	applied := false
 	for cfgPath, update := range updates {
+		// Parse the config contents
```
[the remark this answers is not in the document]
Comment (timothyalexandersoftware) [reply comp/remote-config/rcclient/rcclientimpl/rcclient.go]: deleted in [fbffb0a](https://github.com/DataDog/datadog-agent/pull/24754/commits/fbffb0a537975d05cdb1303171399747a2fab506) 👍

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 6]
```diff
@@ -161,60 +161,77 @@ func (rc rcClient) start() {
...
+		}
+		return
+	}
+
 	applied := false
 	for cfgPath, update := range updates {
 		mrfUpdate, err := parseMultiRegionFailoverConfig(update.Config)
 		if err != nil {
-			pkglog.Errorf("MRF update unmarshal failed: %s", err)
+			pkglog.Errorf("Multi-region Failover update unmarshal failed: %s", err)
```
Comment (tobz) [inline comp/remote-config/rcclient/rcclientimpl/rcclient.go:187]: The smallest of small nits: either go with `Multi-region failover` or `Multi-Region Failover`.

**Review by ameske — APPROVED** [review]

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 7]
```diff
@@ -161,60 +161,77 @@ func (rc rcClient) start() {
...
-							State: state.ApplyStateError,
-							Error: err.Error(),
-						})
-					}
+
+			if mrfUpdate.FailoverMetrics != nil {
+				err = rc.applyMRFRuntimeSetting("multi_region_failover.failover_metrics", *mrfUpdate.FailoverMetrics, cfgPath, applyStateCallback)
+				if err != nil {
+					continue
 				}
```
Comment (tobz) [inline comp/remote-config/rcclient/rcclientimpl/rcclient.go:210]: This is the pedant in me, but could we have an info-level log here that says something like...

> Received remote update for Multi-Region Failover configuration: enabling failover for metrics

where `enabling` is just coming from something like:

```go
state := "disabling"
if *mrfUpdate.FailoverMetrics {
    state = "enabling"
}
```

and then change the log message in `applyMRFRuntimeSetting` to a debug-level log so there's a raw indication, during debugging, of the precise setting name being updated.

Basically, we have a lot of logging for MRF already that is very opaque, and I want to shape it more so that it's useful from a customer perspective: they shouldn't need to know the specific configuration setting to know metrics/logs failover is enabled, we can just tell them in plain English.

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 8]
```diff
@@ -161,60 +161,77 @@ func (rc rcClient) start() {
...
+			})
 			continue
 		}
-		if mrfUpdate != nil && mrfUpdate.Failover != nil {
+
+		if mrfUpdate != nil && (mrfUpdate.FailoverMetrics != nil || mrfUpdate.FailoverLogs != nil) {
+			// If we've received multiple config files updating the failover settings, we should disregard all but the first update and log it, as this is unexpected
 			if applied {
-				pkglog.Infof("Multiple MRF updates received, disregarding update of `multi_region_failover.failover_metrics`/`multi_region_failover.failover_logs` to %t", *mrfUpdate.Failover)
+				pkglog.Infof("Multiple multi-region failover updates received, disregarding update of `multi_region_failover.failover_metrics` to %v and `multi_region_failover.failover_logs` to %v", mrfUpdate.FailoverMetrics, mrfUpdate.FailoverLogs)
```
[the remark this answers is not in the document]
Comment (timothyalexandersoftware) [reply comp/remote-config/rcclient/rcclientimpl/rcclient.go]: The former. This is a technically possible scenario with the RC platform in general, but not one we would expect based on how the product should be publishing these configurations. It should not be common, and as such I bumped it to a warning in [f40e72c](https://github.com/DataDog/datadog-agent/pull/24754/commits/f40e72cc7eeff84a2d980211787166e08681795e) 👍

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go (hunk 6 above)
Comment (timothyalexandersoftware) [reply comp/remote-config/rcclient/rcclientimpl/rcclient.go]: Good call. Took a pass at comments and logs and updated everything to be upper case to be consistent with your previous changes. Updated in [f40e72c](https://github.com/DataDog/datadog-agent/pull/24754/commits/f40e72cc7eeff84a2d980211787166e08681795e)

**Review by tobz — CHANGES_REQUESTED** [review]

**Review by tobz — APPROVED** [review]

File: comp/remote-config/rcclient/rcclientimpl/rcclient.go [hunk 9]
```diff
@@ -161,60 +161,87 @@ func (rc rcClient) start() {
...
-			if !applyError {
-				applied = true
-				applyStateCallback(cfgPath, state.ApplyStatus{State: state.ApplyStateAcknowledged})
+			if mrfUpdate.FailoverLogs != nil {
+				err = rc.applyMRFRuntimeSetting("multi_region_failover.failover_logs", *mrfUpdate.FailoverLogs, cfgPath, applyStateCallback)
+				if err != nil {
+					continue
+				}
+				change := "disabled"
+				if *mrfUpdate.FailoverMetrics {
```
[the remark this answers is not in the document]
Comment (timothyalexandersoftware) [reply comp/remote-config/rcclient/rcclientimpl/rcclient.go]: fixed in [9748f10](https://github.com/DataDog/datadog-agent/pull/24754/commits/9748f103b2c2d6190e51c396bad55d9d2c6849a5) 😅
